File: trtl3-golang-sdk/objects.go
package trtl3sdk

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
)

// POST /objects/upload
// =======================================

func (c *Client) UploadObject(
	bucket string,
	object io.Reader,
	objectName string,
) (bool, error) {
	endpoint := fmt.Sprintf(
		"%s/objects/upload?bucket=%s&object=%s",
		c.url,
		bucket,
		objectName,
	)

	var buffer bytes.Buffer
	writer := multipart.NewWriter(&buffer)

	formFile, err := writer.CreateFormFile("object", filepath.Base(objectName))
	if err != nil {
		return false, fmt.Errorf("Error creating form file: %w", err)
	}

	if _, err := io.Copy(formFile, object); err != nil {
		return false, fmt.Errorf("Error copying object content: %w", err)
	}

	if err := writer.Close(); err != nil {
		return false, fmt.Errorf("Error closing writer: %w", err)
	}

	req, err := http.NewRequest(http.MethodPost, endpoint, &buffer)
	if err != nil {
		return false, fmt.Errorf("Error creating request: %w", err)
	}

	req.Header.Set("Content-Type", writer.FormDataContentType())
	req.Header.Set("Authorization", c.token)

	res, err := c.httpClient.Do(req)
	if err != nil {
		return false, fmt.Errorf("Error while doing request: %w", err)
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusCreated {
		return false, fmt.Errorf("Upload failed with status (%d)", res.StatusCode)
	}

	return true, nil
}

func (c *Client) UploadObjectByPath(
	bucket string,
	path string,
	objectName string,
) (bool, error) {
	file, err := os.Open(path)
	if err != nil {
		return false, fmt.Errorf("Error while opening file: %w", err)
	}
	defer file.Close()

	return c.UploadObject(bucket, file, objectName)
}

func (c *Client) ListObjects(bucketName string) ([]string, error) {
	url := fmt.Sprintf("%s/objects?bucket=%s", c.url, bucketName)

	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("Error trying to create the request: %w", err)
	}

  req.Header.Set("Authorization", c.token)

	res, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("Error while doing a request to the server: %w", err)
	}

	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
    return nil, fmt.Errorf("Failed trying to find created objects(status: %d)", res.StatusCode)
	}

 	bodyBytes, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, fmt.Errorf("Error reading response body: %w", err)
	}

  objects := []string{}

  err = json.Unmarshal(bodyBytes, &objects)
  if err != nil {
    return nil, fmt.Errorf("Error when deserialize response %w", err)
  }

  return objects, nil

}

// =======================================

File: trtl3-node-sdk/lib/objects.ts
import type { AxiosInstance } from 'axios';
import FormData from 'form-data';
import * as fs from 'fs';

export class ObjectApi {
	constructor(private client: AxiosInstance) {}

	async upload(
		bucketName: string,
		objectName: string,
		data: Buffer,
	): Promise<boolean> {
		try {
			const form = new FormData();

			form.append('content', data, { filename: objectName });

			const res = await this.client.post('/objects/upload', form, {
				params: {
					bucket: bucketName,
					object: objectName,
				},
				headers: form.getHeaders(),
			});

			if (res.status !== 201) {
				return false;
			}

			return true;
		} catch (err: unknown) {
			return false;
		}
	}

	async uploadByPath(
		bucketName: string,
		objectName: string,
		path: string,
	): Promise<boolean> {
		try {
			const data = fs.readFileSync(path);

			const uploaded = await this.upload(bucketName, objectName, data);

			return uploaded;
		} catch (err: unknown) {
			return false;
		}
	}

	async list(bucketName: string): Promise<string[]> {
		try {
			const res = await this.client.get('/objects', {
				params: {
					bucket: bucketName,
				},
			});

			if (res.status != 200) {
				return [];
			}

			return res.data;
		} catch (err: unknown) {
			return [];
		}
	}

	async remove(bucketName: string, objectName: string): Promise<boolean> {
		try {
			const res = await this.client.delete('/objects', {
				params: {
					bucket: bucketName,
					object: objectName,
				},
			});

			if (res.status != 200) {
				return false;
			}

			return true;
		} catch (err: unknown) {
			return false;
		}
	}

	async download(
		bucketName: string,
		objectName: string,
	): Promise<Buffer | null> {
		try {
			const res = await this.client.get('/objects', {
				params: {
					bucket: bucketName,
					object: objectName,
				},
			});

			if (res.status !== 200) {
				return null;
			}

			return Buffer.from(res.data);
		} catch (err: unknown) {
			return null;
		}
	}
}
